### DISCUSSION TITLE
BeagleBone Black and GPIO lines when using the Debian kernel

### DISCUSSION THREAD

**Discussion by chombourger** [MentorEmbedded/mtda#196] [General]
We could switch to the Debian kernel for the BeagleBone Black with  #195
Most device drivers appear to work fine. 

GPIO lines however appear to be exposed differently.

I had the signal line of my 5V relay connected to GPIO_20 (P9) 
That line appears to be already ON when the board is booted and toggling of the GPIO does not seem to work

**Comment by vj-kumar** [comment]
Maybe the GPIO is used differently in that kernel version? A dump of the dtb could give an idea. Or better if we find the original dts.

**Reply by vj-kumar** [reply]
Another point being, are we loading the right dtb. /usr/lib/linux-image-5.10.0-10-armmp/am335x-boneblack.dtb

**Reply by chombourger** [reply]
I will add boot logs but yes that DTB is indeed loaded (and not another bbb variant)

**Reply by vj-kumar** [reply]
https://packages.debian.org/buster/amd64/gpiod/filelist

Maybe the gpioinfo command might help here.
https://github.com/torvalds/linux/commit/aafd897a5ac4cb7f9b4f99acc5365a7df1f77aa0

Bullseye is 5.10 kernel I guess, so hopefully this patch should be there.

**Reply by chombourger** [reply]
It's even 5.15 since we are getting the kernel from bpo
I could see the names using gpioinfo and toggle them with gpioset but it did not seem to toggle the line I was expecting

**Reply by chombourger** [reply]
https://github.com/beagleboard/linux has the following gpio-ranges for gpio0:
```
gpio-ranges = <0x20 0x00 0x52 0x08 0x20 0x08 0x34 0x04 0x20 0x0c 0x5e 0x04 0x20 0x10 0x47 0x02 0x20 0x12 0x87 0x01 0x20 0x13 0x6c 0x02 0x20 0x15 0x49 0x01 0x20 0x16 0x08 0x02 0x20 0x1a 0x0a 0x02 0x20 0x1c 0x4a 0x01 0x20 0x1d 0x51 0x01 0x20 0x1e 0x1c 0x02>;
```
while Debian has:
```
gpio-ranges = <0x23 0x00 0x52 0x08 0x23 0x08 0x34 0x04 0x23 0x0c 0x5e 0x04 0x23 0x10 0x47 0x02 0x23 0x12 0x87 0x01 0x23 0x13 0x6c 0x02 0x23 0x15 0x49 0x01 0x23 0x16 0x08 0x02 0x23 0x1a 0x0a 0x02 0x23 0x1c 0x4a 0x01 0x23 0x1d 0x51 0x01 0x23 0x1e 0x1c 0x02>;
```

**Reply by vj-kumar** [reply]
@chombourger I see that the direction is by default input and value is high. Once I set the direction to out I do get value as 0 by default.

**Reply by chombourger** [reply]
good catch: power/gpio does not seem to change the direction to output

**Reply by vj-kumar** [reply]
The default value of high might be due to the internal pull-up resistor. Maybe pin configuration is done in such a way that it is pull-up by default.

**Reply by vj-kumar** [reply]
So, the input comes with pull-up(Value 1)

**Reply by vj-kumar** [reply]
It does set the direction in probe. https://github.com/siemens/mtda/blob/016ae3d0278cc30ce090bb9bc4d9902cda883482/mtda/power/gpio.py#L51

**Comment by vj-kumar** [comment]
Its number 28. Just verified with a multi meter. I am able to toggle it.

P9_12 is GPIO_60 (number 28 for gpio export)

Ref: https://beagleboard.org/static/images/cape-headers.png
```
root@mtda:/sys/class/gpio/gpio28# gpioinfo | grep P9_12
        line  28:      "P9_12"      "sysfs"  output  active-high [used]
root@mtda:/sys/class/gpio/gpio28# 

root@mtda:/sys/class/gpio/gpio28# cat direction 
in
root@mtda:/sys/class/gpio/gpio28# echo out > direction
root@mtda:/sys/class/gpio/gpio28# cat direction 
out
root@mtda:/sys/class/gpio/gpio28# cat value 
0
root@mtda:/sys/class/gpio/gpio28# echo 1 > value 
root@mtda:/sys/class/gpio/gpio28# cat value 
1
```

**Reply by vj-kumar** [reply]
```
root@mtda:/sys/class/gpio/gpio28# dpkg -l | grep linux-image
ii  linux-image-5.15.0-0.bpo.2-armmp     5.15.5-2~bpo11+1               armhf        Linux 5.15 for ARMv7 multiplatform compatible SoCs
ii  linux-image-armmp                    5.15.5-2~bpo11+1               armhf        Linux for ARMv7 multiplatform compatible SoCs (meta-package)
root@mtda:/sys/class/gpio/gpio28# uname -a
Linux mtda 5.15.0-0.bpo.2-armmp #1 SMP Debian 5.15.5-2~bpo11+1 (2022-01-02) armv7l GNU/Linux

```

**Reply by vj-kumar** [reply]
Note to self: VDD_5V on P9_5 / P9_6 is only available when external power supply is connected.

**Reply by chombourger** [reply]
> Its number 28. Just verified with a multi meter. I am able to toggle it.
> 
> P9_12 is GPIO_60 (number 28 for gpio export)
> 
> Ref: https://beagleboard.org/static/images/cape-headers.png
> 
> ```
> root@mtda:/sys/class/gpio/gpio28# gpioinfo | grep P9_12
>         line  28:      "P9_12"      "sysfs"  output  active-high [used]
> root@mtda:/sys/class/gpio/gpio28# 
> 
> root@mtda:/sys/class/gpio/gpio28# cat direction 
> in
> root@mtda:/sys/class/gpio/gpio28# echo out > direction
> root@mtda:/sys/class/gpio/gpio28# cat direction 
> out
> root@mtda:/sys/class/gpio/gpio28# cat value 
> 0
> root@mtda:/sys/class/gpio/gpio28# echo 1 > value 
> root@mtda:/sys/class/gpio/gpio28# cat value 
> 1
> ```

checked on my side and all is good with gpio28
thanks for the help!
